### PR TITLE
Switch from using FirebaseAuth currentUser to db

### LINKED ISSUE #135: Signing in with Apple, photo url becomes null
The `FirebaseUser` object from `FirebaseAuth.instance` has null entries after signing in with Apple when an account with the same email already exists (and is auto linked, see: #134)
- so the strange behaviour seems to be (on the Firebase Auth backend): auto link the accounts and set the photoURL and display name to null. 

So using Sign In With Apple creates and (if email already exists) auto links an account - if linking it seems to set the FirebaseAuth account info to have null displayName and photoURL (is the behaviour just that the details of the account being linked overwrite the current FirebaseUser details?) 

As of #140 we are using the user entries in the database rather than the `FirebaseAuth.instance` data for profile / avatar values.  

#### Problems that still need to be solved: 
- [x] the user can be signed in but the database hasn't yet been updated due to long cloud function cold starts (created #227 to track) 
- [x] auto linking shouldn't be happening (created #226 to track) 
- [x] the `FirebaseAuth.instance` should not be overwritten with null, but marked complete as:
  - this doesn't seem to be up to us 
  - if the behaviour continues we should just make a bug report on Firebase

### PR DESCRIPTION
**Pull request by nickmeinhold**
workaround for #135 

- wrapped the MaterialApp widget in a Provider widget that makes
DatabaseService available in all routes
- added a member currentUserId to DatabaseService, set on auth changes
- used DatabaseService to get user info in place of FirebaseAuth
- added extension to convert a database document snapshot to a UserItem

The PR touches a fair bit of the code base sorry, conceptually the changes are just: 
- get the user info from our database rather than the `FirebaseAuth` instance 
- keep a `currentUserId` in the `DatabaseService`, which is set whenever `onAuthStateChanged` fires an event
- make the `DatabaseService` available to all routes by putting the `Provider` widget above the `MaterialApp` widget 

Known issues:
- Doesn't handle empty db entry, tracked in #147 
- No indication of error in AppBar if user stream emits error #150 
- DatabaseService makes unnecessary calls to Firestore #151

### CODE REVIEW

File: lib/conversations_page.dart [hunk 1]
```diff
@@ -14,13 +14,13 @@ class ConversationsPage extends StatelessWidget {
...
+        leading: StreamBuilder<UserItem>(
+            stream: db.getCurrentUserStream(),
             builder: (context, snapshot) {
-              if (snapshot.hasData) {
-                return UserAvatar(url: snapshot.data.photoURL);
-              } else {
+              if (!snapshot.hasData) {
                 return CircularProgressIndicator();
+              } else {
+                return UserAvatar(url: snapshot.data?.photoURL);
```
Comment (nickmeinhold) [inline lib/conversations_page.dart]: will data always be non-null?

Comment (Jei) [reply lib/conversations_page.dart]: Yes, I left it by mistake to handle error cases, but the builder already returns a CircularProgressIndicator if the stream returns errors or null values. Fixing.

**Comment by nickmeinhold** [comment]
@Jei I ended up just replacing the stream creation and stream.last call with a future by adding a `getCurrentUserFuture` function. 

One thing that *might* be worth addressing at some stage is that we are making unnecessary calls to the firestore as we don't keep the user info. I think we could keep the user info (id, displayName and photoURL) in the DatabaseService, by having the DatabaseService listen to the firestore at `users/{userId}` and keep the user info up to date.  Then we could have the `getCurrentUserFuture` function return a Future that completes immediately if we have local data. I've added #151 to track.

**Comment by Jei** [comment]
> @Jei I ended up just replacing the stream creation and stream.last call with a future by adding a `getCurrentUserFuture` function.

Thanks!

> One thing that _might_ be worth addressing at some stage is that we are making unnecessary calls to the firestore as we don't keep the user info. I think we could keep the user info (id, displayName and photoURL) in the DatabaseService, by having the DatabaseService listen to the firestore at `users/{userId}` and keep the user info up to date. Then we could have the `getCurrentUserFuture` function return a Future that completes immediately if we have local data. I've added #151 to track.

I'm probably thinking a bit too far ahead, but we could do this as a part of a larger "offline mode" feature for the app. What do you think?

**Review by Jei — APPROVED** [review]

**Comment by nickmeinhold** [comment]
Thanks @Jei! That sounds great!  As I'm sure you're aware, we get an "online first, then offline mode" by default with the firestore cache but from memory it only happens after 10 seconds of no internet or something.  It would be great to dig into the firestore offline mode and see what it can do for us, which it sounds like you're doing... I'll be really interested to hear what you find!

**Pull request merged**